receiver/pulsarreceiver/pulsar_receiver.go

```diff
@@ -0,0 +1,325 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulsarreceiver
+
+import (
+	"context"
+	"errors"
+
+	"github.com/apache/pulsar-client-go/pulsar"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/zap"
+)
+
+var errUnrecognizedEncoding = errors.New("unrecognized encoding")
+
+type PulsarTracesConsumer struct {
+	id              config.ComponentID
+	tracesConsumer  consumer.Traces
+	topic           string
+	client          pulsar.Client
+	cancel          context.CancelFunc
+	consumer        pulsar.Consumer
+	ready           chan bool
+	unmarshaler     TracesUnmarshaler
+	settings        component.ReceiverCreateSettings
+	consumerOptions pulsar.ConsumerOptions
+}
+
+func newTracesReceiver(config Config, set component.ReceiverCreateSettings, unmarshalers map[string]TracesUnmarshaler, nextConsumer consumer.Traces) (*PulsarTracesConsumer, error) {
+	unmarshaler := unmarshalers[config.Encoding]
+	if nil == unmarshaler {
+		return nil, errUnrecognizedEncoding
+	}
+
+	options, err := config.ClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	client, err := pulsar.NewClient(options)
+	if err != nil {
+		return nil, err
+	}
+
+	consumerOptions, err := config.ConsumerOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	return &PulsarTracesConsumer{
+		id:              config.ID(),
+		tracesConsumer:  nextConsumer,
+		topic:           config.Topic,
+		unmarshaler:     unmarshaler,
+		settings:        set,
+		client:          client,
+		ready:           make(chan bool),
+		consumerOptions: consumerOptions,
+	}, nil
+}
+
+func (c *PulsarTracesConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancel = cancel
+
+	go func() {
+		err := consumerTracesLoop(ctx, c)
+		if err != nil {
+			c.settings.Logger.Error("Consume tracesConsumer failed", zap.Error(err))
+		}
+	}()
+
+	<-c.ready
+	return nil
+}
+
+func consumerTracesLoop(ctx context.Context, c *PulsarTracesConsumer) error {
+	unmarshaler := c.unmarshaler
+	traceConsumer := c.tracesConsumer
+	close(c.ready)
+
+	subscribe, err := c.client.Subscribe(c.consumerOptions)
+	if nil != err {
+		return err
+	}
+
+	c.consumer = subscribe
+
+	for true {
+		message, err := subscribe.Receive(ctx)
+		if err != nil {
+			return err
+		}
```

Comment:
If the error is not the PulsarClient closed or closing exception, we should continue to pull messages from the consumer.